starlette/middleware/servertiming.py

```diff
@@ -0,0 +1,121 @@
+import threading
+import time
+import typing
+
+from starlette.middleware.base import BaseHTTPMiddleware, RequestResponseEndpoint
+from starlette.requests import Request
+from starlette.types import ASGIApp
+
+
+class Ticker:
+    def __init__(self, name="app", description="main app"):
+        self.name = name
+        self.description = description
+        self._start = self._end = None
+
+    def start(self):
+        self._start = time.time()
+        _add_ticker(self)
+
+    def end(self):
+        self._end = time.time()
+
+    @property
+    def duration(self):
+        if self._start:
+            return int(((self._end or time.time()) - self._start) * 1000)
+        else:
+            return 0
+
+
+class ThreadAwareTickers(threading.local):
+    def __init__(self) -> None:
+        self.__dict__["tickers"] = []
+
+    def add(self, ticker: Ticker) -> None:
+        self.__dict__["tickers"].append(ticker)
+
+    def all(self) -> typing.List[Ticker]:
+        return self.__dict__.setdefault("tickers", [])
+
+    def clear(self) -> None:
+        self.__dict__["tickers"] = []
+
+
+_thread_local = ThreadAwareTickers()
+
+
+def get_tickers() -> typing.List[Ticker]:
+    return _thread_local.all()
+
+
+def discard_all_tickers() -> None:
+    _thread_local.clear()
+
+
+def _add_ticker(ticker) -> None:
+    _thread_local.add(ticker)
+
+
+def ticker_wrapper(func: typing.Callable):
+    def wrapper(*args: typing.Any, **kwargs: typing.Any) -> typing.Any:
+
+        ticker = Ticker(func.__name__)
+        ticker.start()
+
+        data = func(*args, **kwargs)
+        ticker.end()
+
+        return data
+
+    return wrapper
+
+
+class ServerTiming(BaseHTTPMiddleware):
+    """
+    Adds Server-Timing to every HTTP requests
+    """
+
+    def __init__(self, app: ASGIApp, allow_origins: typing.Sequence[str] = ()) -> None:
+        self._allow_origins = allow_origins
+        super().__init__(app, dispatch=self.dispatch)
+
+    async def dispatch(self, request: Request, call_next: RequestResponseEndpoint):
+
+        defaultTimer = Ticker()
+        defaultTimer.start()
```

Comment:
I'm not sure why you're using a threadlocal here, can't you store the time as a float here?

Reply:
Actually, what I'm trying to achieve is every function which have a ticker, will append the threadlocal, so we have a list of tickers, which I can send via server-timing header.

Ex. I have the endpoint /app which contains a query to database, a serialization.

the total time to complete /app is lets say 3 seconds, can be split into a more detailed list.
One can decorate the query to database and put a ticker around a serialization method, thus we will end up with a detailed view:
app - 3 sec
query - 2 sec
serialization - 1 sec

Where I'm stuck is the functions inside the endpoint are executed asynchronously, is there a way to get share the elapsed time of each tickers ?